MazerPlatformer/Character.cs

```diff
@@ -103,30 +110,30 @@ public void MoveInDirection(CharacterDirection direction, GameTime dt)
                     X += MoveByStep();
                     break;
                 default:
-                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
+                    return new InvalidDirectionFailure(direction);
             }
-            SetCharacterDirection(direction);
+            return SetCharacterDirection(direction);
         }
 
-        private void SetState(CharacterStates state)
+        private Either<IFailure, Unit> SetState(CharacterStates state)
         {
             CurrentState = state;
-            OnStateChanged?.Invoke(state);
+            return Ensure( () => OnStateChanged?.Invoke(state));
         }
 
-        private void SetCharacterDirection(CharacterDirection direction)
-        {
-            CurrentDirection = direction;
-
-            SetAnimationDirection(direction);
-            OnDirectionChanged?.Invoke(direction);
-            SetState(CharacterStates.Moving);
+        private Either<IFailure, Unit> SetCharacterDirection(CharacterDirection direction) 
+            => SetAnimationDirection(direction)
+                .Iter(unit => 
+            {
+                CurrentDirection = direction;
```

Comment:
Setting of state of the underlying object is an eye-sore as ideally we don't want to modify the underlying class through its currently unavoidable as this is how the functionality underlying mechanism presents itself.
These segments of the pipeline, it the user transforming function, are also best if they are pure.